### PR TITLE
#336 Configurable Reprompting

### PR DESCRIPTION
**Pull request by andymikulski**
- Adds `repeatOption` and `repeatEvery` properties to `show-heartbeat` actions.
  - `repeatOption` select input consists of three options: `once`, `nag`, or `xdays`
    - `once` will display a prompt once and only once, ever
    - `nag` will display until the user interacts with the prompt in a meaningful way
    - `xdays` will show after waiting at least _x_ days since the last time it appeared
  - If `xdays` is selected, an additional form field for `repeatEvery` appears, which dictates how often the prompt is due to show.
- Adds tests for new show-heartbeat configurations

### CODE REVIEW

File: recipe-server/client/actions/show-heartbeat/index.js [hunk 1]
```diff
@@ -16,35 +19,218 @@ export default class ShowHeartbeatAction extends Action {
...
+
+  /**
+   * Checks the repeat argument for this recipe,
+   * then determines if the recipe can be qualified as 'ran'.
+   * This ultimately decides if the prompt is shown at all
+   * to the end user.
+   *
+   * @return {boolean}        Has the heartbeat been shown?
+   */
+  async heartbeatHasRan() {
```
Comment (andymikulski) [inline recipe-server/client/actions/show-heartbeat/index.js]: I'm confused. This function checks the recipe to see if it's fulfilled the requirements to be considered 'shown' (that `once` has been shown once, that `nag` has been interacted with, etc). It doesn't determine if anything _should_ run - it just checks if the heartbeat has met its specific criteria for being 'ran'.

The function `shouldNotExecute` (Line 214) determines if anything should run (or specifically, shouldn't). It checks normandy's testing state, the global heartbeat timer, and then checks if this specific heartbeat has ran (via `heartbeatHasRan`).

Perhaps the naming convention needs to be changed for some of these functions?

File: recipe-server/client/actions/tests/test_show-heartbeat.js [hunk 2]
```diff
@@ -82,112 +56,300 @@ describe('ShowHeartbeatAction', () => {
...
+        arguments: {
+          repeatOption: 'xdays',
+          repeatEvery,
+        },
+      };
+
+      it('should not show if less than `repeatEvery` days has elapsed', async () => {
+        const action = new ShowHeartbeatAction(normandy, recipeFactory({ ...xdaysConfig }));
+
+        let idx = 0;
```
Comment (andymikulski) [inline recipe-server/client/actions/tests/test_show-heartbeat.js]: Old habit - using `for (let i=0; ...` creates a new variable on each loop, defining it beforehand will only use the one instance. I'll change it.

File: recipe-server/client/control/components/RecipeForm.js [hunk 3]
```diff
@@ -54,15 +55,26 @@ export class RecipeForm extends React.Component {
...
+    const displayedRecipe = this.props.recipe;
+
+    if (!isCloning || !displayedRecipe) {
+      return null;
+    }
+
+    this.cloneCache = this.cloneCache || {};
+    const cache = this.cloneCache;
+
+    if (!cache[displayedRecipe.id]) {
```
Comment (andymikulski) [inline recipe-server/client/control/components/RecipeForm.js]: I had noticed that I had set up the cloning message to be rendered via 
```{ this.renderCloningMessage() }```, meaning the function fires on each render. Adding the caching at least saved some cycles, but it's not a big deal if it's removed.

File: recipe-server/client/actions/show-heartbeat/index.js (hunk 1 above)
Comment (mythmon) [inline recipe-server/client/actions/show-heartbeat/index.js]: Sorry, I was a bit confused here. It isn't clear to me what "run" means in this context. I was expecting that "run" meant either (1) the action has ever been executed by a runner for this recipe, or if (2) a heartbeat prompt has ever been shown for this recipe.

I think that you are describing run as (3) this recipe has fulfilled all of it's criteria for never being executed again.

The comment "Has the heartbeat ever been shown" makes me think (2). Overall I find this flow pretty hard to follow, since it is split up among a lot of different functions, and there are a lot of ill defined concepts (like "run").

All that is to say: I think changing the comment in the `@return` section will be enough to ward off enough future confusion here.

**Comment by mythmon** [comment]
Looks good. Sorry about the confusion with naming.

File: recipe-server/client/actions/show-heartbeat/index.js (hunk 1 above)
Comment (andymikulski) [inline recipe-server/client/actions/show-heartbeat/index.js]: I rewrote a bunch of the language and ended up renaming the function all together - https://gist.github.com/andymikulski/f2531bdab2bec531ffc0e188f1b62fc2 - Let me know if it looks good and I'll push the changes before merging the branch.

Comment (mythmon) [inline recipe-server/client/actions/show-heartbeat/index.js]: +1, that reads a lot clearer to me. Thanks for re-working it!
